source/extensions/filters/network/mysql_proxy/mysql_terminal_filter.cc

```diff
@@ -0,0 +1,522 @@
+    // auth switch.
+    auth_method = AuthMethod::NativePassword;
+  }
+  seed = greet.getAuthPluginData();
+  auto login = MessageHelper::encodeClientLogin(
+      auth_method, parent.upstream_username_, parent.upstream_password_, parent.connect_db_, seed);
+  login.setSeq(greet.getSeq() + 1);
+
+  parent.sendRemote(login);
+  parent.stepRemoteSession(greet.getSeq() + 2, MySQLSession::State::ChallengeResp41);
+}
+
+void MySQLTerminalFilter::UpstreamEventHandler::onClientLogin(ClientLogin&) {
+  ENVOY_LOG(error, "mysql filter: upstream onClientLogin impossible callback is called");
+}
+
+void MySQLTerminalFilter::UpstreamEventHandler::onAuthSucc() {
+  parent.downstream_event_handler_->onBothAuthSucc();
+  ready = true;
+  parent.stepRemoteSession(MYSQL_RESPONSE_PKT_NUM, MySQLSession::State::ReqResp);
+  if (parent.downstream_event_handler_->buffer.length()) {
+    parent.downstream_event_handler_->decoder->onData(parent.downstream_event_handler_->buffer);
+  }
+}
+
+void MySQLTerminalFilter::UpstreamEventHandler::onClientLoginResponse(
+    ClientLoginResponse& login_resp) {
+  parent.onClientLoginResponse(login_resp);
+
+  absl::optional<ErrMessage> err;
+
+  switch (login_resp.getRespCode()) {
+  case MYSQL_RESP_AUTH_SWITCH: {
+    auto& auth_switch = dynamic_cast<AuthSwitchMessage&>(login_resp);
+    if (!auth_switch.isOldAuthSwitch() &&
+        auth_switch.getAuthPluginName() != "mysql_native_password") {
+      ENVOY_LOG(info, "auth plugin {} is not supported", auth_switch.getAuthPluginName());
+      err = MessageHelper::defaultError(
+          "mysql proxy is failed to auth to upstream due to auth plugin is not support (now "
+          "support mysql_native_password, mysql_old_password)");
+      break;
+    }
+    auto auth_method =
+        auth_switch.isOldAuthSwitch() ? AuthMethod::OldPassword : AuthMethod::NativePassword;
+    seed = auth_switch.getAuthPluginData();
+
+    auto switch_resp =
+        MessageHelper::encodeClientLogin(auth_method, parent.upstream_username_,
+                                         parent.upstream_password_, parent.connect_db_, seed);
+    switch_resp.setSeq(login_resp.getSeq() + 1);
+    parent.sendRemote(switch_resp);
+    parent.stepRemoteSession(login_resp.getSeq() + 2, MySQLSession::State::AuthSwitchMore);
+    break;
+  }
+  case MYSQL_RESP_OK:
+    onAuthSucc();
+    break;
+  case MYSQL_RESP_ERR:
+    ENVOY_LOG(info, "mysql proxy failed of auth, info {}",
+              dynamic_cast<ErrMessage&>(login_resp).getErrorMessage());
+    err = MessageHelper::defaultError(
+        fmt::format("mysql proxy is failed to auth upstream database server due to {}",
+                    dynamic_cast<ErrMessage&>(login_resp).getErrorMessage()));
+    break;
+  case MYSQL_RESP_MORE:
+  default:
+    ENVOY_LOG(info, "unhandled message resp code {}", login_resp.getRespCode());
+    err = MessageHelper::defaultError(fmt::format("mysql proxy is failed to auth upstream database "
```

Comment:
mysql authentication with upstream server failed due to unhandled message resp code {}